### PR TITLE
Shopify OAuth provider - OWIN integration / implementation

### PR DESCRIPTION
**Pull request by jsinh**
This one was a bit different as in, the OAuth URLs where not fixed. The base URL sub-domain (shop-name) keeps changing based on the choice of user wants to authorize against the desired shop.

Updated the demo sample a little without disturbing other OAuth usage, to demonstrate the usage.
Hope it meets the expectation for the PR to be accepted and get released via NuGet !!

### CODE REVIEW

**Comment by jsinh** [comment]
ping @jerriep :grimacing:

**Comment by jerriep** [comment]
Jip, sorry. Was going to look at it on Friday, but then some other things came up. Will try and fit it in today...

**Comment by jerriep** [comment]
@jsinh Can you please rather add the ShopName as a property to the `ShopifyAuthenticationOptions` class and pass it along as a parameter to `UseShopifyAuthentication()`. 

Since it is a required parameter, please ensure to do null checking and throw a `ArgumentNullException` if it was not passed in...

Requiring the user to add it to the ChallengeResult is not obvious, and also not the right place to add it IMHO. Once you have done that please remove all the code you added to change the ChallengeResult from the demo project.

File: Owin.Security.Providers/Shopify/ShopifyAuthenticationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            if (null == app)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+
+            if (null == options)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
```
Comment (jerriep) [inline Owin.Security.Providers/Shopify/ShopifyAuthenticationExtensions.cs:24]: Please remove C# 6 `nameof`. This is not a c# 6 project

**Comment by jsinh** [comment]
@jerriep cannot set it in `ShopifyAuthenticationOptions` and pass it as parameter to `UseShopifyAuthentication()`.

These things are set at the start of the application and / or only once to initiate this OAuth provider and put in the pipeline to be used.

The `ShopName` keep changing for each user who is trying to get authorized or create account using Shopify as option. In shopify they do not have a common domain OAuth URL(s). They have provided OAuth isolated API for each shop that users or shop owners can create so the `ShopName` keeps changing every time any new shop owner wants to Authorize and Authenticate against Shopify Shop they wish to do it against.

One clear real world usage and the one that I am interested in is: Shopify APP Store. Apps listed there use these OAuth to Authorized Shop Owner's and Shop and install their APP in the Shop based on it to access Shop's resources and provide services.

Hope it makes sense.
Also removed the `nameof`, sorry about that.

Jsinh

**Comment by jerriep** [comment]
Oh wow, that is not good.....

Well look, either way you will need to please write a blog post which we can link to in order to tell people how to use this, as it is not obvious. Would you be able to do that please?

I am just going to sleep on this one for a day or so and think if there is not another way to do this. But I do understand your dilemma...

**Comment by jerriep** [comment]
Regarding the blog post - it may actually be better if you could write a guide for http://www.oauthforaspnet.com/ as that is the central place where all the other guides are located...

Whichever way you feel most comfortable with

**Comment by jsinh** [comment]
Fair enough, I understand this one is a bit weird and odd. It took me more than a sleep to get my head clear and found this to be the best approach IMO.

Regarding the blog post. I can do both on my blog and at oauthforaspnet. The only problem is I do not have my Jekyll working well yet so may take sometime to figure out what is so wrong with it. Once that is fixed I will do the oauthforaspnet.

Have a nice sleep, I will try to get that sorted before you wake up.

**Comment by jerriep** [comment]
Yeah Jekyll can be tricky. I found Portable Jekyll the easiest to use:
https://github.com/madhur/PortableJekyll

Simply download the latest release and extract to a folder on your harddrive. Then whenever you want to use it, execute `setpath.cmd` to set the correct paths and after that you should be able to run the Jekyll commands.

**Comment by jsinh** [comment]
PortableJekyll **bookmarked** :smile: thanks!
Submitted PR for post on How to use Shopify OAuth Provider.

**Pull request merged**